presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/TransformCorrelatedNonAggregationScalarToJoin.java

```diff
@@ -0,0 +1,181 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner.iterative.rule;
+
+import com.facebook.presto.matching.Captures;
+import com.facebook.presto.matching.Pattern;
+import com.facebook.presto.metadata.FunctionRegistry;
+import com.facebook.presto.spi.type.BooleanType;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.iterative.Lookup;
+import com.facebook.presto.sql.planner.iterative.Rule;
+import com.facebook.presto.sql.planner.optimizations.PlanNodeDecorrelator;
+import com.facebook.presto.sql.planner.optimizations.PlanNodeDecorrelator.DecorrelatedNode;
+import com.facebook.presto.sql.planner.optimizations.ScalarSubqueryToJoinRewriter;
+import com.facebook.presto.sql.planner.optimizations.ScalarSubqueryToJoinRewriter.SubqueryEquivalentJoin;
+import com.facebook.presto.sql.planner.plan.EnforceSingleRowNode;
+import com.facebook.presto.sql.planner.plan.FilterNode;
+import com.facebook.presto.sql.planner.plan.LateralJoinNode;
+import com.facebook.presto.sql.planner.plan.MarkDistinctNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.ProjectNode;
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.SimpleCaseExpression;
+import com.facebook.presto.sql.tree.StringLiteral;
+import com.facebook.presto.sql.tree.WhenClause;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.matching.Pattern.nonEmpty;
+import static com.facebook.presto.spi.StandardErrorCode.SUBQUERY_MULTIPLE_ROWS;
+import static com.facebook.presto.spi.type.StandardTypes.BOOLEAN;
+import static com.facebook.presto.sql.planner.optimizations.PlanNodeSearcher.searchFrom;
+import static com.facebook.presto.sql.planner.optimizations.QueryCardinalityUtil.isScalar;
+import static com.facebook.presto.sql.planner.plan.Patterns.LateralJoin.correlation;
+import static com.facebook.presto.sql.planner.plan.Patterns.lateralJoin;
+import static com.facebook.presto.sql.tree.BooleanLiteral.TRUE_LITERAL;
+import static com.facebook.presto.util.MorePredicates.isInstanceOfAny;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Scalar filter scan query is something like:
+ * <pre>
+ *     SELECT a,b,c FROM rel WHERE a = correlated1 AND b = correlated2
+ * </pre>
+ * <p>
+ * This optimizer can rewrite to aggregation over a left outer join:
+ * <p>
+ * From:
+ * <pre>
+ * - LateralJoin (with correlation list: [C])
```

Comment:
What if somebody specifically written lateral join and expects more than one result row from subquery?

Shouldn't scalar subqueries be represented by `ApplyNode`? Then you know that you expect at most one row from subquery.

Then the whole functionality could be expressed by two rules:
```
1. 
ApplyNode(correlation)
 - input
 - subquery

rewrite to:

LateralJoin[LEFT](correlation)
 - input
 - subquery

```
this rewrite fires only when `QueryCardinalityUtil#isAtMostScalar(subquery)==true`

the second rule
```
2.
ApplyNode(correlation)
 - input
 - subquery

rewrite to:

Filter(CASE isDistinct WHEN true THEN true ELSE fail('Scalar sub-query has returned multiple rows'))
  MarkDistinct(isDistinct)
    LateralJoin[LEFT](correlation)
     - AssignUniqueId
       - input
     - subquery

```
this rewrite fires when `1)` did not fire or `QueryCardinalityUtil#isAtMostScalar(subquery)==false`

`LateralJoin` rewrite would be taken care by other rules.
Am I missing something?

Reply:
> What if somebody specifically written lateral join and expects more than one result row from subquery?

Sure, it should work. Here, subquery in lateral will be wrapped with `EnsureSingleRow`.

> Shouldn't scalar subqueries be represented by ApplyNode? Then you know that you expect at most one row from subquery.

I thought it works that way. Anyway, I still know it is scalar subquery as it is wrapped with `EnforceSingleRow`

What do you propose sounds reasonable, but this is kind of refactor which in my opinion exceeds the scope of this PR.

Reply:
> Here, subquery in lateral will be wrapped with `EnsureSingleRow`.

Still those two proposed rewrites apply. You just check if child of `EnsureSingleRow` is `isAtMostScalar`. If so, then taking out `EnsureSingleRow` should be enough. Otherwise you rewrite like in 2) (without `EnsureSingleRow` of course)

Reply:
We don't need to rewrite directly to join, do we? We can rewrite to lateral join:
```
LateralJoin(correlation, type)
 - input
 - subquery

rewrite to:

Filter(CASE isDistinct WHEN true THEN true ELSE fail('Scalar sub-query has returned multiple rows'))
  MarkDistinct(isDistinct)
    LateralJoin[type](correlation)
     - AssignUniqueId
       - input
     - subquery
```